bcda/auth/backend.go

```diff
@@ -23,17 +33,38 @@ func init() {
 // Hash is a cryptographically hashed string
 type Hash string
 
-// NewHash creates a hashed string from a source string, return it as a new Hash value.  While this is not
-// intended for production password hashing, as a matter of principle we intend to add a salt and implement
-// bcrypt or another hash appropriate for storing secrets in ticket BCDA-1130.
-func NewHash(source string) Hash {
-	sum := sha256.Sum256([]byte(source))
-	return Hash(fmt.Sprintf("%x", sum))
+// NewHash creates a hashed string from a source string, return it as a new Hash value.
+func NewHash(source string) (Hash, error) {
+	salt := make([]byte, saltSize)
+	_, err := rand.Read(salt)
+	if err != nil {
+		return Hash(""), err
+	}
+
+	h := pbkdf2.Key([]byte(source), salt, hashIter, hashKeyLen, sha512.New)
+	return Hash(fmt.Sprintf("%s:%s", base64.StdEncoding.EncodeToString(salt), base64.StdEncoding.EncodeToString(h))), nil
 }
 
 // IsHashOf accepts an unhashed string, which it first hashes and then compares to itself
 func (h Hash) IsHashOf(source string) bool {
-	return h == NewHash(source)
+	// Avoid comparing with an empty source so that a hash of an empty string is never successful
+	if source == "" {
```

Comment:
I get that it is invalid and unhelpful to hash "", but what would be the harm in just hashing ""?

Reply:
Because it's being used to verify credentials, this is the last line of defense in case some weird error mode allowed a hashed "" value to be saved as the client secret.